math/miller_rabin.cpp

```diff
@@ -0,0 +1,185 @@
+ * @param modulo
+ * @return the modular exponentiation of a^(reverse_binary) % n
+ */
+int modular_exponentiation(int a, std::vector<int> rev_binary, int n) {
+    if (n == 1)
+        return 0;
+    int b = 1;
+    if (rev_binary.size() == 0)
+        return b;
+    int A = a;
+    if (rev_binary[0] == 1)
+        b = a;
+
+    for (std::vector<int>::iterator it = rev_binary.begin() + 1;
+         it != rev_binary.end(); ++it) {
+        A = A * A % n;
+        if (*it == 1)
+            b = A * b % n;
+    }
+    return b;
+}
+
+/**
+ * Function to raise a number to a power
+ * @param base
+ * @param power
+ * @return base^power
+ */
+int power(int num, int pow) {
+    if (pow == 1) {
+        return num;
+    }
+    return num * power(num, pow - 1);
+}
+
+/** Function for testing the conditions that are satisfied when a number is
+ * prime.
+ * 	@param number d such that d * 2 ^ r = num - 1 for some r >= 1
+ * 	@param number being tested.
+ * 	@return false if n is composite and true if n is probably prime.
+ */
+bool miller_test(int d, int num) {
+    // random number seed
+    std::random_device rd_seed;
+    // random number generator
+    std::mt19937 gen(rd_seed());
+    // Uniformly distributed range [2, num - 2] for random numbers
+    std::uniform_int_distribution<> distribution(2, num - 2);
+    // Random number generated in the range [2, num -2].
+    int random = distribution(gen);
+    // vector for reverse binary of the power
+    std::vector<int> power = reverse_binary(d);
+    // x = random ^ d % num
+    int x = modular_exponentiation(random, power, num);
+    // miller conditions
+    if (x == 1 || x == num - 1) {
+        return true;
+    }
+
+    while (d != num - 1) {
+        x = (x * x) % num;
+        d *= 2;
+        if (x == 1) {
+            return false;
+        }
+        if (x == num - 1) {
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Function that test (probabilistically) whether a given number is a prime
+ * based on the Miller-Rabin Primality Test.
+ * @param number to be tested
+ * @param number of repetitions for the test.
+ * @return false if num is composite and true if its probably prime
+ *
+ * First we check whether the num input is less than 4, if so we can determine
+ * whther this is a prime or composite by checking for 2 and 3.
+ * Next we check whether this num is odd (as all primes greater than 2 are odd).
+ * Next we write our num in the following format num = 2^r * d + 1. After
+ * finding r and d for our input num, we use for loop repeat number of times
+ * inside which we check the miller conditions using the function miller_test.
+ * If miller_test returns false then the number is composite
+ * After the loop finishes completely without issueing a false return call,
+ * we can conclude that this number is probably prime.
+ */
+bool miller_rabin_primality_test(unsigned int num, unsigned int repeats) {
```

Comment:
is this function also valid for `unsigned long`, `unsigned short`, `unsigned long long`? then maybe this function can be templated

Reply:
@kvedala I have tried to fix all the things that you mentioned here, could you please take a look?